plugins/hls-refactor-plugin/test/Main.hs

```diff
@@ -2131,14 +2136,423 @@ insertNewDefinitionTests = testGroup "insert new definition actions"
+  , testSession "simple" $ do
+      let foo =
+            [ "module Foo where"
+            , ""
+            , "bar = 1"
+            , ""
+            , "foo True = _select [True]"
+            , "  where"
+            , "foo False = False"
+            ]
+          foo' =
+            [ "module Foo where"
+            , ""
+            , "bar = 1"
+            , ""
+            , "foo True = _select [True]"
+            , "  where"
+            , "    _select = _"
+            , "foo False = False"
+            ]
+      docB <- createDoc "ModuleB.hs" "haskell" (T.unlines foo)
+      _ <- waitForDiagnostics
+      InR action@CodeAction { _title = actionTitle } : _
+                    <- filter (\(InR CodeAction{_title=x}) -> "Add to " `isPrefixOf` T.unpack x ) <$>
+                     getCodeActions docB (R 4 0 4 50)
+      liftIO $ actionTitle @?= "Add to where ‘_select’"
+      executeCodeAction action
+      contentAfterAction <- documentContents docB
+      liftIO $ contentAfterAction @?= T.unlines foo'
+  , testSession "empty where" $ do
+      let foo =
+            [ "module Foo where"
+            , ""
+            , "bar = 1"
+            , ""
+            , "foo True = _select [True]"
+            , "  where"
+            , "foo False = False"
+            ]
+          foo' =
+            [ "module Foo where"
+            , ""
+            , "bar = 1"
+            , ""
+            , "foo True = _select [True]"
+            , "  where"
+            , "    _select = _"
+            , "foo False = False"
+            ]
+      docB <- createDoc "ModuleB.hs" "haskell" (T.unlines foo)
+      _ <- waitForDiagnostics
+      InR action@CodeAction { _title = actionTitle } : _
+                    <- filter (\(InR CodeAction{_title=x}) -> "Add to " `isPrefixOf` T.unpack x ) <$>
+                     getCodeActions docB (R 4 0 4 50)
+      liftIO $ actionTitle @?= "Add to where ‘_select’"
+      executeCodeAction action
+      contentAfterAction <- documentContents docB
+      liftIO $ contentAfterAction @?= T.unlines foo'
+  , testSession "untyped error" $ do
+      let foo =
+            ["module Foo where"
+            ,""
+            ,"foo = select"
+            ]
+          foo' =
+            ["module Foo where"
+            ,""
+            ,"foo = select"
+            , "  where"
+            , "    select = _"
+            ]
+      docB <- createDoc "ModuleB.hs" "haskell" (T.unlines foo)
+      _ <- waitForDiagnostics
+      InR action@CodeAction { _title = actionTitle } : _
+                    <- filter (\(InR CodeAction{_title=x}) -> "Add to " `isPrefixOf` T.unpack x ) <$>
+                     getCodeActions docB (R 2 0 2 50)
+      liftIO $ actionTitle @?= "Add to where ‘select’"
+      executeCodeAction action
+      contentAfterAction <- documentContents docB
+      liftIO $ contentAfterAction @?= T.unlines foo'
+  , testSession "untyped error" $ do
+      let foo =
+            [ "module Foo where"
```

Comment:
Comments are tough. 

Which version of ghc-exactprint?

Reply:
@santiweight you should be using the high-level API as demonstrated in the various `addLocalDeclX` tests from https://github.com/alanz/ghc-exactprint/blob/master/tests/Test/Transform.hs#L318

Reply:
And for current ghc-exactprint master, I reproduced your example with a changer test using

```hs
addLocaLDecl7 :: Changer
addLocaLDecl7 libdir lp = do
  let
      doAddLocal = do
        decls0 <- hsDecls lp
        replaceDecls lp decls0

  (lp',_,_w) <- runTransformT doAddLocal
  debugM $ "log:[\n" ++ intercalate "\n" _w ++ "]log end\n"
  return lp'
```

And test file

```hs
module AddLocalDecl7 where

foo = 1
  where
    -- hi
    baz = 1
```

Reply:
Thanks for these examples @alanz

I realize now that the newer versions of ghc-exactprint are not available to me until 9.4.x is supported in HLS...

I am using 9.2.4 for the refactoring actions I have been working on up until now, which I believe corresponds to ghc-exactprint 1.6.

I am thinking it might be best to hold out for newer versions of ghc-exactprint to be available with 9.4 and then use newer versions of the API then... Alternatively, I might try again to use the exact-ring available in HLS and see if the results can be better than what I've been producing before I saw these examples in the ghc-exactprint repo!

Reply:
9.4 _is_ supported in HLS